imxrt-hal/src/adc.rs

```diff
@@ -223,6 +223,205 @@ where
+/// static ADC_BUFFER2: hal::dma::Buffer<[u16; 256]> = hal::dma::Buffer::new([0; 256]);
+/// static mut streaming_adc: Option<hal::adc::StreamingAdc<hal::iomuxc::adc::ADC1, t41::P24>> = None;
+///
+/// #[interrupt]
+/// fn DMA8_DMA24() {
+///     unsafe {
+///         if let Some(adc) = &mut streaming_adc {
+///             adc.handle_interrupt();
+///         }
+///     }
+/// }
+///
+/// ...
+///
+/// let (adc1_builder, _) = adc.clock(&mut ccm.handle);
+/// let mut adc1 = adc1_builder.build(
+///     adc::ClockSelect::default(),
+///     adc::ClockDivision::default()
+/// );
+/// let mut a1 = adc::AnalogInput::new(pins.p24);
+///
+/// let mut channel = dma_channels[8].take().unwrap();
+/// channel.set_interrupt_on_completion(true);
+/// unsafe {
+///     streaming_adc = Some(hal::adc::StreamingAdc::new(
+///         adc1,
+///         a1,
+///         channel,
+///         &ADC_BUFFER,
+///         &ADC_BUFFER2
+///     ));
+///     cortex_m::peripheral::NVIC::unmask(interrupt::DMA8_DMA24);
+/// }
+///
+/// ...
+///
+/// // Inside the main loop
+/// let mut sadc = unsafe { streaming_adc.as_mut().unwrap() };
+/// sadc.poll(|buf| {
+///     adc_min = *buf.iter().min().unwrap();
+///     adc_max = *buf.iter().max().unwrap();
+/// });
+/// ```
+pub struct StreamingAdc<ADCx, P> {
+    buffer0: &'static crate::dma::Buffer<[u16; 256]>,
+    buffer1: &'static crate::dma::Buffer<[u16; 256]>,
+    peripheral: crate::dma::Peripheral<
+        AdcSource<ADCx, P>,
+        u16,
+        crate::dma::Linear<u16>,
+        crate::dma::Linear<u16>,
+    >,
+    unhandled_buffer_count: core::sync::atomic::AtomicUsize,
+    off_buffer: crate::dma::Linear<u16>,
+    _pin: PhantomData<P>,
+}
+
+impl<ADCx, P> StreamingAdc<ADCx, P>
+where
+    ADCx: adc::ADC + AdcDmaSource,
+    P: Pin<ADCx>,
+{
+    /// Create a new StreamingAdc object. The returned object will not begin
+    /// generating samples until start() is called.
+    pub fn new(
+        adc: ADC<ADCx>,
+        pin: AnalogInput<ADCx, P>,
+        channel: crate::dma::Channel,
+        buffer0: &'static crate::dma::Buffer<[u16; 256]>,
+        buffer1: &'static crate::dma::Buffer<[u16; 256]>,
+    ) -> Self {
+        let mut off_buffer = crate::dma::Linear::new(&buffer0).unwrap();
+        off_buffer.set_transfer_len(256);
+        let mut peripheral = crate::dma::receive_u16(AdcSource::new(adc, pin), channel);
+
+        Self {
+            buffer0,
+            buffer1,
+            peripheral,
+            unhandled_buffer_count: core::sync::atomic::AtomicUsize::new(0),
+            off_buffer,
+            _pin: PhantomData,
+        }
+    }
+
+    /// Begin transferring samples from the ADC. Once samples are flowing, the
+    /// user must call poll() frequently enough to keep up with the generated
+    /// samples, or else the interrupt handler will panic.
+    pub fn start(&mut self) {
+        let mut buffer = crate::dma::Linear::new(&self.buffer1).unwrap();
```

Comment:
> I'm afraid that calling start() twice, where buffer1 refers to a previously-used Buffer, will result in a panic. Have we seen this in practice?

If `start()` should only be called once, we could document that, or express the behavior with typestates.

Reply:
Yeah, so far I've only used it letting it run continuously for eternity, so starting hasn't had to happen twice.

I think a `Circular` would be the way to go, I dislike how the double-buffering has to restart the DMA buffer periodically. That's how I was originally doing it, but I was having trouble getting `Circular` to notice that the DMA engine added elements to it. Is there some sort of poll() method I need to call, or should it in principle happen automatically?

A `impl Iterator<Item = u16>` would work fine, it could be the `Drain` iterator, worst case the user just stack-allocates a buffer for whatever they need. The only alternative I can think of would be `Fn(&[u16], &[u16]) -> usize`, returning the number of elements consumed, but that's a bit more complicated (and the second argument would usually be an empty slice anyway).

Reply:
>  I was having trouble getting Circular to notice that the DMA engine added elements to it.

Sorry for the trouble! We're calling [`reserve()`](https://docs.rs/imxrt-hal/0.4.2/imxrt_hal/dma/struct.Circular.html#method.reserve) on the circular buffer before submitting it to the channel? That should signal how much data we want to receive in the next DMA transfer. After that, it should be automatic. See the [`memcpy` example](https://github.com/mciantyre/teensy4-rs/blob/73433b3ae6cbd7754729d0dcecca61bbfe0419a2/examples/bsp/src/bin/dma_memcpy.rs#L73) in the Teensy 4 BSP for a demo.